developer_guides/firmware/llext_modules.rst

```diff
@@ -0,0 +1,78 @@
+With it modules are built as shared or relocatable ELF objects with an addition
+of a cryptographic signature and a manifest. When loaded and instantiated,
+Zephyr LLEXT functionality is used to dynamically resolve module internal as
+well as SOF external code and data references. In the future support for
+inter-module linking will be added.
+
+Accessing the base firmware from LLEXT modules
+**********************************************
+
+LLEXT modules can access all code and data from the base firmware exported,
+using the ``EXPORT_SYMBOL()`` macro. Therefore writing LLEXT modules isn't very
+different from built-in ones.
+
+Implementing LLEXT modules
+**************************
+
+At the moment only modules, implementing the Module Adapter API
+:ref:`apps-comp-world` are supported.
+
+As explained above, LLEXT modules in general look very similar to native SOF
+code, with the only restriction of having no access to not-exported symbols. If
+any such symbols are needed to LLEXT modules, additionally exporting them can be
+considered.
+
+LLEXT modules should also contain a ``.buildinfo`` section, containing a
+``struct sof_module_api_build_info`` object and a ``.module`` section,
+containing a ``struct sof_man_module_manifest`` object. The latter should also
+contain a pointer to a module entry point function, returning a pointer to the
+module's ``struct module_interface`` instance. See
+src/samples/audio/smart_amp_test_ipc4.c for an example.
+
+A TOML configuration file is needed for building of LLEXT modules too. It is
+generated by the C preprocessor at build time from the same components, as would
+be used for a monolithic build. A small additional file to bring components
+together is needed too, similar to
+src/samples/audio/smart_amp_test_llext/llext.toml.h.
+
+Finally an additional CMakeLists.txt is needed similar to
+src/samples/audio/smart_amp_test_llext/CMakeLists.txt. Currently it contains a
+lot of open-coded instructions, later it will be ported to use Zephyr's
+``add_llext_target()`` function.
+
+With that in place, it is also possible to switch between monolithic and modular
+builds by specifying the module as "tristate" in its Kconfig and selecting "m"
+for modular builds, however, for modules, exclusively built as LLEXT objects,
+this might not be necessary. Thus it is possible to implement third party Module
+Adapter drivers, that would be built exclusively as loadable modules.
+
+Note, that a single LLEXT object file can also contain multiple Module Adapter
+modules. In that case an array of ``struct sof_module_api_build_info`` objects
+is needed and the TOML file should contain those multiple module entries too.
+
+Installation
+************
+
+As described in :ref:`intel_debug_introduction` the |SOF| Linux kernel driver
+loads SOF modules, using their UUIDs. For SOF internal modules the process
+of creation and installation of modules in a deployment tree is automated by the
+xtensa-build-zephyr.py script.
```

Comment:
Is xtensa-build-zephyr.py optional? If yes, is there a way to build _only_ a module or a plan to support that? If not, how do you build with a different key?

If it is required to build everything at the same time with the same key then what's the purpose of LLEXT?

Reply:
@marc-hb the purpose is not to have to load everything into SRAM at once. Only currently active pipelines need all modules in SRAM, the rest can be kept out one way or another.
xtensa-build-zephyr.py is optional in the sense, that it is "only" calling other programs to actually manipulate files, it isn't doing any deep binary manipulation itself. As you have seen in https://github.com/thesofproject/sof/pull/9013/commits/f85900353ce226a8093971f8bc91babd5debf2ba it "just" locates all the modules, built by cmake, calls rimage to sign them and installs the result into a deployment tree, including soft links. Yes, you can do that all manually, but it isn't very convenient or simple.
As for building only a module - you need a full Zephyr+SOF build environment - you need .config. It should be possible to sign modules with alternative keys, just call rimage manually as mentioned above.

Reply:
> It should be possible to sign modules with alternative keys, just call rimage manually as mentioned above.
> ...
> Yes, you can do that all manually, but it isn't very convenient or simple.

Please mention at least that so readers have at least some vague idea of what is "automated" and some high level idea of what problem xtensa-build-zephyr.py solves.